### PR TITLE
`HTTPClient`: added signature validation and introduced `ErrorCode.signatureVerificationFailed`

### PR DESCRIPTION
**Pull request by NachoSoto**
Fixes [CSDK-634].

## Changes:
- Created `HTTPResponseValidationResult`
- Added `HTTPResponseValidationResult` and `HTTPClient.ResponseHeaders` to `HTTPResponse`
- Added new `ErrorCode.signatureVerificationFailed`
- Extracted `SigningType` to create `MockSigning` to test `HTTPClient` without bothering with signatures
- Refactored `ETagManager` to use `HTTPResponse`, making it easier to pass through response headers and verification results
- Stored `HTTPResponseValidationResult` in `ETagManager` cache
- `HTTPClient` handles signatures from response based on `SystemInfo.responseVerificationLevel`
- Renamed `HTTPRequest.createIntegrityEnforcedRequestRequest` to `HTTPRequest.createWithResponseVerification`
- Added lots of tests for this new behavior in `HTTPClient`
- Added lots of new tests for `ETagManager`

[CSDK-634]: https://revenuecats.atlassian.net/browse/CSDK-634?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Tests/UnitTests/Networking/ETagManagerTests.swift [hunk 1]
```diff
@@ -212,17 +286,214 @@ class ETagManagerTests: TestCase {
...
+            statusCode: .success,
+            data: actualResponse,
+            validationResult: .notRequested
+        ).asData()
+
+        let response = self.eTagManager.eTagHeader(for: request)
+        expect(response[ETagManager.eTagHeaderName]) == eTag
+    }
+
+    func testCachedResponseWithNoValidationResultIsIgnored() {
```
Comment (NachoSoto) [inline Tests/UnitTests/Networking/ETagManagerTests.swift:380]: This tests backwards compatibility with old responses. We handle that as a cache miss.

File: Tests/UnitTests/Networking/HTTPClientTests.swift [hunk 2]
```diff
@@ -96,26 +106,6 @@ class HTTPClientTests: TestCase {
         expect(headers?.keys).toNot(contain(HTTPClient.nonceHeaderName))
     }
 
-    @available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.2, *)
-    func testRequestIncludesRandomNonce() throws {
```
Comment (NachoSoto) [inline Tests/UnitTests/Networking/HTTPClientTests.swift:100]: This test was moved below to `SignatureVerificationHTTPClientTests`

File: Sources/Misc/Signing+ResponseValidation.swift [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+//
+//  Created by Nacho Soto on 2/8/23.
+
+import Foundation
+
+extension HTTPResponse where Body == Data {
+
+    static func createAndValidate(with response: HTTPURLResponse,
+                                  body: Data,
+                                  request: HTTPRequest,
```
Comment (bisho) [inline Sources/Misc/Signing+ResponseValidation.swift:20]: Can we extend `HTTPRequest` to have a `IntegrityVerifiedHTTPRequest` that guarantees contains a Nonce? And then make this signature expect that class in particular. Doesn't make sense to validate a `HTTPRequest` that is not sending Nonces.

Comment (NachoSoto) [reply Sources/Misc/Signing+ResponseValidation.swift:20]: That's definitely an alternative approach, but that's not very Swift-y. We'd have to move away from it being a `struct` and lose the value semantics on this type.

I think the main issue is this is bad name. It returns `HTTPResponseValidationResult.notRequested` if the request doesn't need to be verified, so maybe instead of `createAndValidate` it should just be `HTTPResponse.create(with:...)`.

Comment (NachoSoto) [reply Sources/Misc/Signing+ResponseValidation.swift:20]: Alternatively we'd have to make `HTTPResponse` a protocol and use a generic `<Response: HTTPResponseType>` everywhere to maintain type safety at compile time instead of doing that at runtime. But at that point it would be an overkill.

File: Sources/Networking/HTTPClient/HTTPResponse.swift [hunk 4]
```diff
@@ -18,19 +18,48 @@ struct HTTPResponse<Body: HTTPResponseBody> {
...
+                return bodyDescription
+            } else {
+                return "\(type(of: self.body))"
+            }
+        }()
+
+        return """
+        HTTPResponse(" +
+        statusCode: \(self.statusCode.rawValue),
+        body: \(body),
```
Comment (NachoSoto) [inline Sources/Networking/HTTPClient/HTTPResponse.swift:59]: OMG code coverage pointing to a bug here, amazing.

**Comment by NachoSoto** [comment]
Apologies for the big diff 🙏🏻 I couldn't really split the changes here into multiple PRs since they're all connected.
Most of the changes are a lot of new tests though

File: Sources/Misc/Signing+ResponseValidation.swift (hunk 3 above)
Comment (bisho) [reply Sources/Misc/Signing+ResponseValidation.swift]: Hummm... I see... I have no clue about swift :)

And something like createRequestAndValidator() that returns a request with the nonce, and a validator ready to use (contains the nonce)? It feels weird to have this generic create and validate that sometimes will not validate, plus there is not a lot of type safety, there is no way to enforce validation for some kind of request.

That way the request remains generic, you will just build a validator when requesting a request with validation. If not requesting validation, you will just have a create that builds a request.

Again, all this not knowing a thing about swift patterns :)

Comment (NachoSoto) [reply Sources/Misc/Signing+ResponseValidation.swift]: I renamed this, I think it's clearer now. I do get your point of seemingly validating requests that don't require it.
`HTTPResponse` always has an `HTTPResponseValidationResult` (even if it's `.notRequested`).
So `HTTPResponse.create(with:body:request:..)` is the new API. It may or may not do any validation, but it'll return an `HTTPResponse` with the appropriate information. Does that sound less weird?

File: Tests/UnitTests/Networking/ETagManagerTests.swift [hunk 5]
```diff
@@ -212,17 +286,214 @@ class ETagManagerTests: TestCase {
...
+        let cacheKey = url.absoluteString
+
+        let actualResponse = "response".asData
+
+        self.mockUserDefaults.mockValues[cacheKey] = """
+        {
+        "e_tag": "\(eTag)",
+        "status_code": 200,
+        "data": "\(actualResponse.asFetchToken)",
+        "validation_result": 1
```
Comment (thecadams) [inline Tests/UnitTests/Networking/ETagManagerTests.swift:449]: nit: maybe consider `"validation_result": \(.validated.rawValue)`

**Review by thecadams — COMMENTED** [review]
Nice change @NachoSoto. Not a swift expert, but looks good to me. The API with the optional requesting of verification is going to prove handy if we decide to do this for more endpoint in future..

PS particularly liked how simple the `sign` and `createSalt` test methods got, the simple keypair generation with `private let (privateKey, publicKey) = SigningTests.createRandomKey()`, and how packing things into test classes creates clean encapsulation. Learned a bunch about Swift reading this, thanks for sending it through!

File: Sources/Networking/HTTPClient/HTTPClient.swift [hunk 6]
```diff
@@ -221,10 +226,12 @@ private extension HTTPClient {
...
+            .mapToResponse(response: httpURLResponse,
+                           request: request.httpRequest,
+                           signing: self.signing,
+                           verificationLevel: self.systemInfo.responseVerificationLevel)
             .map {
-                self.eTagManager.httpResultFromCacheOrBackend(with: httpURLResponse,
-                                                              data: $0.body,
+                self.eTagManager.httpResultFromCacheOrBackend(with: $0,
                                                               request: urlRequest,
                                                               retried: request.retried)
```
[the remark this answers is not in the document]
Comment (thecadams) [reply Sources/Networking/HTTPClient/HTTPClient.swift:236]: This is based on the empty body case in https://github.com/RevenueCat/purchases-ios/pull/2272/files#diff-2a4a83f4e5d43494fe43c0b623281c54d24bd60193c7de185f37e565b0c72de3R107-R108 and the comment suggesting that code assumes a cache hit in the caller..

File: Tests/UnitTests/Networking/ETagManagerTests.swift (hunk 5 above)
Comment (NachoSoto) [reply Tests/UnitTests/Networking/ETagManagerTests.swift:449]: In this particular case I wanted to hardcore it so we never break backwards compatibility if somebody changes the associated value. I'll add a comment to explain.

File: Sources/Networking/HTTPClient/HTTPClient.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient/HTTPClient.swift:236]: > If there's an empty response and no etag, looks like this will be nil due to nothing in cache. Is that ok for callers?

Yeah that's confusing, it took me a while to wrap my head around why that was done when I refactored this months ago.
The answer to that is in `HTTPClient.handle(urlResponse:request:urlRequest:data:error:)`. If this returns `nil` (the scenario you described), it gets retried. There's plenty of test coverage for that.

> Is that ok for callers?

As for users of `HTTPClient`, ultimately responses can't be `nil` thanks to the strong type safety provided by its API.
The `Completion` type is defined as
```swift
Completion<Value: HTTPResponseBody> = (HTTPResponse<Value>.Result) -> Void
```

Which is basically `(Result<HTTPResponse<Value>, NetworkError>) -> Void`. In other words, there will be a response or an error, but never an empty or `nil` response.

File: Tests/UnitTests/Networking/ETagManagerTests.swift (hunk 5 above)
Comment (NachoSoto) [reply Tests/UnitTests/Networking/ETagManagerTests.swift:449]: Actually I'm adding plenty of test coverage for that serialization in #2277 so I'll follow your suggestion 👍🏻

**Comment by NachoSoto** [comment]
I'm thinking of renaming the `ErrorCode` here instead if we agree with the name in #2277.

File: Sources/Networking/HTTPClient/ETagManager.swift [hunk 7]
```diff
@@ -35,28 +35,27 @@ class ETagManager {
 
     func eTagHeader(for urlRequest: URLRequest, refreshETag: Bool = false) -> [String: String] {
         var storedETag = ""
-        if !refreshETag, let storedETagAndResponse = storedETagAndResponse(for: urlRequest) {
+        if !refreshETag,
+           let storedETagAndResponse = self.storedETagAndResponse(for: urlRequest),
+           storedETagAndResponse.validationResult != .failedValidation {
```
[the remark this answers is not in the document]
Comment (aboedo) [reply Sources/Networking/HTTPClient/ETagManager.swift:40]: also... do we ever clean these up? 

It feels like if we have an eTag but it doesn't have validation information and validation has been activated, we should kill the old eTag, but I don't recall what our cleanup mechanism for this was

[the remark this answers is not in the document]
Comment (aboedo) [reply Sources/Networking/HTTPClient/ETagManager.swift:40]: also, we're not storing when validationResult = failed, so this would never be true, right?

[the remark this answers is not in the document]
Comment (aboedo) [reply Sources/Networking/HTTPClient/ETagManager.swift:40]: last question: what are we doing here when validation wasn't requested at all? Like, if validation is disabled, this should behave exactly as it always has

File: Tests/UnitTests/Mocks/MockHTTPClient.swift [hunk 8]
```diff
@@ -31,7 +31,12 @@ class MockHTTPClient: HTTPClient {
             // swiftlint:disable:next force_try
             let data = try! JSONSerialization.data(withJSONObject: response)
 
-            self.init(response: .success(.init(statusCode: statusCode, body: data)),
+            self.init(response: .success(.init(
+                statusCode: statusCode,
+                responseHeaders: [:],
+                body: data,
+                validationResult: .notRequested
+            )),
```
Comment (tonidero) [inline Tests/UnitTests/Mocks/MockHTTPClient.swift:39]: Hmm indentation here seems hard to read IMO... Maybe extract the `success` to a variable?

File: Sources/Networking/HTTPClient/ETagManager.swift [hunk 9]
```diff
@@ -166,7 +171,24 @@ extension ETagManager.Response {
...
+
+// MARK: -
+
+private extension HTTPResponse {
+
+    var shouldStore: Bool {
+        return (
+            self.statusCode != .notModified &&
+            !self.statusCode.isServerError &&
+            self.validationResult != .failedValidation
```
Comment (tonidero) [inline Sources/Networking/HTTPClient/ETagManager.swift:190]: 🎉

File: Sources/Networking/HTTPClient/HTTPClient.swift [hunk 10]
```diff
@@ -72,6 +76,7 @@ extension HTTPClient {
 
     static let authorizationHeaderName = "Authorization"
     static let nonceHeaderName = "X-Nonce"
+    static let responseSignatureHeaderName = "x-signature"
```
Comment (tonidero) [inline Sources/Networking/HTTPClient/HTTPClient.swift:79]: Should this be 
```suggestion
    static let responseSignatureHeaderName = "X-Signature"
```

**Review by tonidero — COMMENTED** [review]
Still need to review tests but looking good!

File: Sources/Networking/HTTPClient/ETagManager.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient/ETagManager.swift:40]: > we're interchangeably using validation and verification throughout the feature, we should pick one and settle. I think verification is more technically appropriate for this

I fixed those inconsistencies in #2277. I went with `validation` but we can change it back.

> It feels like if we have an eTag but it doesn't have validation information and validation has been activated, we should kill the old eTag

Hmm good point, let me see if we have a test case for that and if not I'll add it.

File: Sources/Networking/HTTPClient/HTTPClient.swift (hunk 10 above)
Comment (NachoSoto) [reply Sources/Networking/HTTPClient/HTTPClient.swift:79]: Hmm good point.

File: Sources/Misc/Signing+ResponseValidation.swift [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        body: Data,
+        headers: HTTPClient.ResponseHeaders,
+        request: HTTPRequest,
+        publicKey: Signing.PublicKey?,
+        signing: SigningType.Type
+    ) -> HTTPResponseValidationResult {
+        guard let nonce = request.nonce,
+              let publicKey = publicKey,
+              #available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.2, *) else {
+            return .notRequested
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Misc/Signing+ResponseValidation.swift:61]: We talked about this offline.
We're going to solve this by making the "validated" APIs only available in iOS 13+, to force users to decide how to handle older versions.

The configuration API is already not available below, so it's also clear that you can't use it on older devices.

File: Tests/UnitTests/Mocks/MockHTTPClient.swift (hunk 8 above)
Comment (NachoSoto) [reply Tests/UnitTests/Mocks/MockHTTPClient.swift:39]: Done.

File: Sources/Networking/HTTPClient/ETagManager.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient/ETagManager.swift]: Okay, updated `ETagManager` to ignore etags for the case where the stored cache had no validation but the new request requires a signature.

**Comment by NachoSoto** [comment]
All comments addressed.

File: Sources/Misc/Signing+ResponseValidation.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply Sources/Misc/Signing+ResponseValidation.swift:61]: So this will change in a followup PR I imagine right?

File: Sources/Networking/HTTPClient/HTTPResponse.swift [hunk 12]
```diff
@@ -18,19 +18,54 @@ struct HTTPResponse<Body: HTTPResponseBody> {
 
     typealias Result = Swift.Result<Self, NetworkError>
 
-    let statusCode: HTTPStatusCode
-    let body: Body
+    var statusCode: HTTPStatusCode
+    var responseHeaders: HTTPClient.ResponseHeaders
+    var body: Body
+    var validationResult: HTTPResponseValidationResult
```
Comment (tonidero) [inline Sources/Networking/HTTPClient/HTTPResponse.swift:24]: I may be missing something but why did we make these `var` instead of `let`?

File: Tests/UnitTests/Mocks/MockETagManager.swift [hunk 13]
```diff
@@ -13,25 +13,39 @@ import Foundation
 // swiftlint:disable type_name
 class MockETagManager: ETagManager {
 
+    struct ETagHeaderRequest {
+        var urlRequest: URLRequest
+        var refreshETag: Bool
+        var signatureVerificationEnabled: Bool
```
Comment (tonidero) [inline Tests/UnitTests/Mocks/MockETagManager.swift:19]: Same here, could these be let? I don't think callers of this mock would need to modify this?

**Review by tonidero — APPROVED** [review]
LGTM!

File: Sources/Networking/HTTPClient/HTTPResponse.swift (hunk 12 above)
Comment (NachoSoto) [reply Sources/Networking/HTTPClient/HTTPResponse.swift:24]: So we can modify the validation result after it's initially created in `HTTPClient` (just an implementation detail).
Note that `struct`s have [value semantics](https://www.swiftbysundell.com/articles/utilizing-value-semantics-in-swift/), so this isn't like having mutable data in other languages.

Example:
```swift
struct Data {
  // Technically mutable
  var a: Int
  var b: String
}

let d = Data(a: 1, b: "test")
d.a +=1 // Error, `d` is immutable!

func f(_ d: Data) {
  d.a += 1 // Error, d is immutable
}
```

Additionally, `struct`s get implicitly copied (they have _value_ semantics, not _reference_), they're passed by _value_, so you can't lead to shared mutable state.

So basically it doesn't make a difference if these are `var`s, it just makes creating copies with different values easier without needing to do it manually calling the constructor again.

File: Tests/UnitTests/Mocks/MockETagManager.swift (hunk 13 above)
Comment (NachoSoto) [reply Tests/UnitTests/Mocks/MockETagManager.swift:19]: Ditto, it doesn't make a difference, it would only be mutable for a `var ETagHeaderRequest` anyway.

File: Sources/Networking/HTTPClient/HTTPResponse.swift (hunk 12 above)
Comment (tonidero) [reply Sources/Networking/HTTPClient/HTTPResponse.swift:24]: Right, I asked about this before and I forgot 😅 thanks for explaining again!
